lms/djangoapps/grades/models.py

```diff
@@ -107,7 +107,8 @@ def hash_from_blockrecords(self, blocks):
             blocks = BlockRecordSet(blocks)
 
         try:
-            model = self.create_from_blockrecords(blocks)
+            with transaction.atomic():
+                model = self.create_from_blockrecords(blocks)
```

Comment:
@nasthagiri I've wrapped the 2 `except IntegrityError:` locations I found in grading code.